File: code/jvm/repository/src/testFixtures/java/pt/isel/repositories/mem/InvitationRepositoryInMem.java
package pt.isel.repositories.mem;

import pt.isel.domain.channels.AccessType;
import pt.isel.domain.channels.Channel;
import pt.isel.domain.invitations.Invitation;
import pt.isel.domain.invitations.InvitationStatus;
import pt.isel.domain.users.UserInfo;
import pt.isel.repositories.invitations.InvitationRepository;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class InvitationRepositoryInMem implements InvitationRepository {
    private final List<Invitation> invitations = new ArrayList<>();
    private long nextId = 1;

    @Override
    public Invitation create(String token, UserInfo createdBy, Channel channel, AccessType accessType, LocalDateTime expiresAt) {
        if (findByToken(token) != null) throw new RuntimeException("invitations_token_key");
        Invitation inv = new Invitation(nextId++, token, createdBy, channel, accessType, expiresAt);
        invitations.add(inv);
        return inv;
    }

    @Override
    public Invitation findByToken(String token) {
        return invitations.stream().filter(i -> i.token().equals(token)).findFirst().orElse(null);
    }

    @Override
    public List<Invitation> findByChannelId(Long channelId) {
        return invitations.stream().filter(i -> i.channel().id().equals(channelId)).toList();
    }

    @Override
    public boolean consume(Long id) {
        for (int i = 0; i < invitations.size(); i++) {
            Invitation inv = invitations.get(i);
            if (inv.id().equals(id) && inv.status() == InvitationStatus.PENDING) {
                invitations.set(i, new Invitation(inv.id(), inv.token(), inv.createdBy(), inv.channel(), inv.accessType(), inv.expiresAt(), InvitationStatus.ACCEPTED));
                return true;
            }
        }
        return false;
    }

    @Override
    public Invitation findById(Long id) {
        return invitations.stream().filter(i -> i.id().equals(id)).findFirst().orElse(null);
    }

    @Override
    public List<Invitation> findAll() {
        return new ArrayList<>(invitations);
    }

    @Override
    public void save(Invitation entity) {
        invitations.removeIf(i -> i.id().equals(entity.id()));
        invitations.add(entity);
    }

    @Override
    public void deleteById(Long id) {
        invitations.removeIf(i -> i.id().equals(id));
    }

    @Override
    public void clear() {
        invitations.clear();
        nextId = 1;
    }
}

File: code/jvm/repository/src/test/kotlin/pt/isel/repositories/mem/InvitationRepositoryInMem.kt
package pt.isel.repositories.mem

import jakarta.inject.Named
import pt.isel.domain.channels.AccessType
import pt.isel.domain.channels.Channel
import pt.isel.domain.invitations.Invitation
import pt.isel.domain.users.UserInfo
import pt.isel.repositories.invitations.InvitationRepository
import java.time.LocalDateTime

/**
 * Naif in memory repository non thread-safe and basic sequential id. Useful for unit tests purpose.
 */
@Named
class InvitationRepositoryInMem : InvitationRepository {
    private val invitations = mutableListOf<Invitation>()

    override fun create(
        token: String,
        createdBy: UserInfo,
        channel: Channel,
        accessType: AccessType,
        expiresAt: LocalDateTime,
    ): Invitation =
        Invitation(
            invitations.size.toLong() + 1,
            token,
            createdBy,
            channel,
            accessType,
            expiresAt,
        ).also { invitations.add(it) }

    override fun findById(id: Long): Invitation? = invitations.firstOrNull { it.id == id }

    override fun findByToken(token: String): Invitation? = invitations.firstOrNull { it.token == token }

    override fun findByChannelId(channelId: Long): List<Invitation> = invitations.filter { it.channel.id == channelId }

    override fun findAll(): List<Invitation> = invitations.toList()

    override fun save(entity: Invitation) {
        invitations.removeIf { it.id == entity.id }.apply { invitations.add(entity) }
    }

    override fun deleteById(id: Long) {
        invitations.removeIf { it.id == id }
    }

    override fun clear() {
        invitations.clear()
    }
}
